uwblueprint/starter-code-v2
# Add default rate limiting

Fixes a CodeQL error in #150 



## Implementation description
* Added default rate limit of 15 calls per minute for TypeScript REST, TypeScript GraphQL, and Python REST
* This should apply to each endpoint separately 
* `app.use(limiter);` is commented for now since it applies to both REST requests and every GraphQL request, so you would only get 15 calls on /graphql overall
* Added some missing error validation to API calls on frontend



## Steps to test
TypeScript GraphQL
1. Go to localhost:3000 and login. 
2. Click Display Entities and then Go Back 15 times (you can change the rate limit for testing if this is too high). On the 16th try, it should rate limit you and stop displaying the results. We currently do not handle errors on this page so you may need to add an onError to the useQuery hook to see the error.

TypeScript REST
1. Set up TypeScript REST on the frontend (change DisplayTableContainer to use REST queries). Also, uncomment `app.use(limiter);` in server.ts.
2. Go to localhost:3000 and login. 
3. Repeat step 2 in TypeScript GraphQL above.

Python REST
1. Set up Python REST - remember to change MG_DATABASE_URL in .env to the python db
2. Go to localhost:3000 and login. 
3. Repeat step 2 in TypeScript GraphQL above.


## What should reviewers focus on?
* What should the default rate limit be?
* Are there any other places in the frontend that I missed with error validation? (with rate limiting, some places where we did not catch errors will now crash more easily)
*  Regarding frontend error validation, one case that happens is that if the user cannot refresh token due to rate limit, success will be false and setAuthenticatedUser(null) will be called, logging out the user - what should the expected behaviour be in this case?


## Checklist
